--- a/programs/ipsec/ipsec.in
+++ b/programs/ipsec/ipsec.in
@@ -130,18 +130,53 @@ ipsec_help() {
     exit 2
 }
 
+# add nflog-all
+nflog_nftables_add() {
+    nft add table ip filter
+    nft add chain ip filter ipsec-all-input "{ type filter hook input priority 0; policy accept; }"
+    nft add chain ip filter ipsec-all-output "{ type filter hook output priority 0; policy accept; }"
+    nft add rule ip filter ipsec-all-input meta ipsec exists log prefix "all-ipsec" group 50
+    nft add rule ip filter ipsec-all-output rt ipsec exists log prefix "all-ipsec" group 50
+}
+
+nflog_nftables_delete() {
+    f=$(nft list chain ip filter ipsec-all-input 2>/dev/null | wc -l)
+    if [ ${f} -gt 0 ]; then
+    	nft delete chain ip filter ipsec-all-input
+    fi
+
+    f=$(nft list chain ip filter ipsec-all-output 2>/dev/null | wc -l)
+    if [ ${f} -gt 0 ]; then
+        nft delete chain ip filter ipsec-all-output
+    fi
+}
+
+nflog_iptables_delete() {
+    local GROUP=$1
+    iptablese -D INPUT  -m policy --dir in  --pol ipsec -j NFLOG --nflog-group "${GROUP}" --nflog-prefix all-ipsec
+    iptablese -D OUTPUT -m policy --dir out --pol ipsec -j NFLOG --nflog-group "${GROUP}" --nflog-prefix all-ipsec
+}
+
 ipsec_stopnflog() {
     NFGROUP=$(ASAN_OPTIONS=detect_leaks=0 "${IPSEC_EXECDIR}/addconn" --ctlsocket "${CTLSOCKET}" --configsetup | grep -v "#" | grep nflog | sed -e "s/^.*=//" -e "s/'//g");
     if [ -z "${NFGROUP}" ]; then
 	exit 0
     fi
-    iptables -D INPUT  -m policy --dir in  --pol ipsec -j NFLOG --nflog-group "${NFGROUP}" --nflog-prefix all-ipsec
-    iptables -D OUTPUT -m policy --dir out --pol ipsec -j NFLOG --nflog-group "${NFGROUP}" --nflog-prefix all-ipsec
+    firewall_cmd
+
+    if [ "${FIREWALL}" = IPTABLES ]; then
+        nflog_iptables_delete "${NFGROUP}"
+    elif [ "${FIREWALL}" = NFTABLES ]; then
+        nflog_nftables_delete
+    else
+        echo "unknown firewall comaand ${FIREWALL_CMD} and ${FIREWALL} expect ipables or nft"
+        exit 1
+    fi
+
     exit 0
 }
 
-ipsec_checknflog() {
-    NFGROUP=$(ASAN_OPTIONS=detect_leaks=0 "${IPSEC_EXECDIR}/addconn" --ctlsocket "${CTLSOCKET}" --configsetup | grep -v "#" | grep nflog| sed -e "s/^.*=//" -e "s/'//g");
+checknflog_iptables() {
     if [ -n "${NFGROUP}" ]; then
 	OLDNFGROUP=$(iptables -L -n | grep "all-ipsec nflog-group" | sed "s/^.* //" | tail -1);
 	if [ -n "${OLDNFGROUP}" ]; then
@@ -152,8 +187,7 @@ ipsec_checknflog() {
 	    else
 		# delete rules with old group number
 		echo "deleting rules with old nflog group ${OLDNFGROUP}"
-		iptables -D INPUT  -m policy --dir in  --pol ipsec -j NFLOG --nflog-group ${OLDNFGROUP} --nflog-prefix all-ipsec
-		iptables -D OUTPUT -m policy --dir out --pol ipsec -j NFLOG --nflog-group ${OLDNFGROUP} --nflog-prefix all-ipsec
+                nflog_iptables_delete "${OLDNFGROUP}"
 	    fi
 	fi
 	# insert rules with current group number
@@ -164,11 +198,64 @@ ipsec_checknflog() {
 	OLDNFGROUP=$(iptables -L -n | grep "all-ipsec nflog-group" | sed "s/^.* //" | tail -1);
 	if [ -n "${OLDNFGROUP}" ]; then
 	    echo "deleting rules with old nflog group ${OLDNFGROUP}"
-	    iptables -D INPUT  -m policy --dir in  --pol ipsec -j NFLOG --nflog-group "${OLDNFGROUP}" --nflog-prefix all-ipsec
-	    iptables -D OUTPUT -m policy --dir out --pol ipsec -j NFLOG --nflog-group "${OLDNFGROUP}" --nflog-prefix all-ipsec
+        nflog_iptables_delete "${OLDNFGROUP}"
 	fi
 	echo "nflog ipsec capture disabled"
     fi
+}
+
+checknflog_nftables() {
+    if [ -n "${NFGROUP}" ]; then
+	f=$(nft list table ip filter 2>/dev/null | wc -l)
+	if [ ${f} -gt 0 ]; then
+	    f=$(nft list chain ip filter ipsec-all-input 2>/dev/null | wc -l)
+	    if [ ${f} -gt 0 ]; then
+		echo "found ipsec-all-input"
+	    else
+		nflog_nftables_add
+	    fi
+	else
+	    nflog_nftables_add
+	fi
+    else
+	nflog_nftables_delete
+    fi
+
+    # nft delete rule filter input handle 15
+    # nft add rule ip filter ipsec-all-input meta ipsec exists ipsec in ip saddr 192.168.1.0/24 accept
+    # nft -a list
+    # nft -a list chain ip filter ipsec-all-output
+    # nft delete chain ip filter ipsec-all-output
+}
+
+firewall_cmd() {
+    HAVE_IPTABLES=true
+    HAVE_NFTABLES=true
+
+    if [ "${HAVE_NFTABLES}" = true ]; then
+        FIREWALL=NFTABLES
+    elif [ "${HAVE_IPTABLES}" = true ]; then
+        FIREWALL=IPTABLES
+    else
+        FIREWALL=""
+    fi
+}
+
+ipsec_checknflog() {
+    NFGROUP=$(ASAN_OPTIONS=detect_leaks=0 "${IPSEC_EXECDIR}/addconn" --ctlsocket "${CTLSOCKET}" --configsetup | grep -v "#" | grep nflog| sed -e "s/^.*=//" -e "s/'//g");
+
+    firewall_cmd
+
+    if [ "${FIREWALL}" = IPTABLES ]; then
+        checknflog_iptables
+    elif [ "${FIREWALL}" = NFTABLES ]; then
+        checknflog_nftables
+    else
+	FIREWALL=""
+	echo "unknown firewall command ${FIREWALL} expect iptables or nft"
+	exit 1
+    fi
+
     exit 0
 }
 
