--- a/src/lxc/log.h
+++ b/src/lxc/log.h
@@ -304,13 +304,13 @@ __lxc_unused static inline void LXC_##LE
  * Helper macro to define errno string.
  */
 #if HAVE_STRERROR_R
-	#ifdef STRERROR_R_CHAR_P
+	#if STRERROR_R_CHAR_P
 	char *strerror_r(int errnum, char *buf, size_t buflen);
 	#else
 	int strerror_r(int errnum, char *buf, size_t buflen);
 	#endif
 
-	#ifdef STRERROR_R_CHAR_P
+	#if STRERROR_R_CHAR_P
 		#define lxc_log_strerror_r                                               \
 			char errno_buf[PATH_MAX / 2] = {"Failed to get errno string"};   \
 			char *ptr = NULL;                                                \
